mysys/CMakeLists.txt

```diff
@@ -92,6 +92,12 @@ ELSEIF(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64|AARCH64")
   IF(CMAKE_COMPILER_IS_GNUCC)
     include(CheckCXXSourceCompiles)
 
+    CHECK_CXX_SOURCE_COMPILES("
+    // print error 'NEON intrinsics not available with the soft-float ABI.
+    // Please use -mfloat-abi=softfp or -mfloat-abi=hard'
+    #include <arm_neon.h>
+    int main(){}" HAVE_ARMV8_NEON)
```

Comment:
As far as I understand, the intention of this pull request is to fix builds that target ARMv5 or ARMv6. On my Raspberry Pi 2, `${CMAKE_SYSTEM_PROCESSOR}` evaluates to `armv7l`. I would assume it to be something similar on the targeted environments. It definitely should not reach this branch of the CMake code. We currently do not seem to have any CMake code that is specific to 32-bit ARM architectures.

I see that NEON is optional in ARMv8-R. But, which compilation problem on ARMv8 a.k.a. Aarch64 is this solving? I would guess that an ARMv8 build environment would have this header file available, and we’d have to perform a run-time check, in case the code is executing on a processor where NEON is not available. Only the file `mysys/crc32/crc32_arm64.c` (for ARMv8, not 32-bit ARMs) should depend on this.

Reply:
`HWCAP_NEON` (bit 12) is defined in the 32bit userspace, but in arm64 bit 12 is reused for a different capability.

Also these bits mean different things on 32bit userspace:
```
  #ifndef HWCAP_CRC32
  # define HWCAP_CRC32 (1 << 7)
  #endif
  
  #ifndef HWCAP_PMULL
  # define HWCAP_PMULL (1 << 4)
  #endif
```

So as the build system is building crc32/crc32_arm64.c on 32 bit, its going to be broken.

Until Debian start doing a cross-compulation (documented - https://wiki.debian.org/CrossCompiling and [tested](http://crossqa.debian.net/build/mariadb_1:10.11.2-3_armhf_20230423012238.log)) rather than salsa hacky, the packages generated, and the tests in the same hacked environment, aren't worth the BTUs of power that made them. Setting an env  BUILD_ARCH: 'armhf' ([ref](https://salsa.debian.org/salsa-ci-team/pipeline/raw/master/salsa-ci.yml)) is insufficient.

Setting env HOST_ARCH=armhf (or armel) appears to be the necessary bit to trigger [cross compulation](https://salsa.debian.org/salsa-ci-team/pipeline/raw/master/salsa-ci.yml).

Reply:
As this is mainly soft-float problem which NEON solves.  I didn't know that event ARMv8 NEON can be optional I though it's mandatory. I've try to follow that cross compile to produce same binaries.

Reply:
This branch of the CMakeLists.txt is under a `CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64|AARCH64"` condition, it shouldn't be reached on armv[67] as all armv[67] are 32 bit. Otto's pipeline shows crc32_arm64.cc compiled on armhf and armel and even has `SYSTEM processor: aarch64` in the libmariadb side. This is clearly wrong. The tests passing on a 64bit hardware with a userspace don't prove its right.

So problem 1 - ensure 32 bit builds don't hit this branch.

on https://buildd.debian.org/status/package.php?p=mariadb - it appears that the real build doesn't compile this, so the salsa build is different from the real. So small relief - though ' SYSTEM processor: armv8l' is in both the armhf and armel output.

Problem 2 -  ARMv8 without neon.

The `arm_neon.h` header in the mysys/crc32_arm64.c seems there to purely define `vmull_p64` as `return (poly128_t) __builtin_arm_crypto_vmullp64 ((uint64_t) __a, (uint64_t) __b);` this appears to be a [pmull](https://developer.arm.com/documentation/ddi0597/2023-12/SIMD-FP-Instructions/VMULL--integer-and-polynomial---Vector-Multiply-Long--integer-and-polynomial--?lang=en) feature. This is already checked.

There is no neon features used directly.

option 1 - remove the header and use the `__builtin_arm_crypto_vmullp64` directly.

option 2 - i think is what you've done, try to make it compile with the header. regardless of if the neon features are used.

I tried https://salsa.debian.org/grooverdan/mariadb-server/-/pipelines/623885 yesterday.


So for solving Problem 2, I'm ok with this patch, with the one reset of `SET(CMAKE_REQUIRED_FLAGS)` below, and a much more focused commit message on the problem 2 thing that its solving, including why we even need to play with NEON at all. A comment in the CMakeList.txt around this would also be good.

Reply:
branch back port also, its still a bug fix.

Reply:
What would be lowest branch to use? 10.11 or 10.4?

Reply:
10.5 where mysys/crc32/crc32_arm64.c was added.

Reply:
Now rebased